csrc/dynamic_transform.cpp

```diff
@@ -385,6 +541,222 @@ void DynamicTransformConcretizer::concretize() {
+    auto tv = info_->initialInfo()->lookUpTV(empty_tv_descr.tv_name);
+    auto rfactor = TensorDomain::noReductions(tv->getMaybeRFactorDomain());
+    std::vector<Val*> new_shape;
+    new_shape.reserve(rfactor.size());
+    for (auto id : rfactor) {
+      new_shape.push_back(id->getMaybeExpandedExtent());
+    }
+    for (auto ax : empty_tv_descr.empty_axes) {
+      // Hard-code zero extent for empty axes. This lets us detect empty input
+      // and output tensors during scheduling/execution.
+      registerConcretization(new_shape[ax], fusion->zeroVal());
+      new_shape[ax] = fusion->zeroVal();
+    }
+
+    auto hasEmptyRootReductionAxis = [&empty_tv_descr](TensorView* out_tv) {
+      return std::any_of(
+          empty_tv_descr.empty_axes.begin(),
+          empty_tv_descr.empty_axes.end(),
+          [&out_tv](size_t ax) {
+            return out_tv->getRootDomain().at(ax)->isReduction();
+          });
+    };
+
+    // Given a TensorView, get a shape with hard-coded zeroes
+    auto orig_shape = [](TensorView* out_tv) -> std::vector<Val*> {
+      auto nored_axes =
+          TensorDomain::noReductions(out_tv->getMaybeRFactorDomain());
+      // Output shape is simply the same as the original reduction. If there
+      // were zeros in the non-Reduction axes, it would be replaced by
+      // full() directly.
+      std::vector<Val*> out_shape(nored_axes.size());
+      std::transform(
+          nored_axes.begin(),
+          nored_axes.end(),
+          out_shape.begin(),
+          [](IterDomain* id) -> Val* { return id->getMaybeExpandedExtent(); });
+      return out_shape;
+    };
+
+    std::unordered_map<TensorView*, TensorView*> replaced;
+    auto maybeReplaced = [&replaced](TensorView* tv) -> TensorView* {
+      auto it = replaced.find(tv);
+      if (it == replaced.end()) {
+        return tv;
+      }
+      return it->second;
+    };
+
+    // Replace uses whose outputs might not be empty. Many expressions are
+    // guaranteed to have empty outputs if any of the inputs are empty; for
+    // example simple unary or binary ops. In those cases, we don't need to
+    // doctor the Fusion since they will have an empty tensor downstream which
+    // will cut off their dependence, resulting in those uses becoming dead
+    // code.
+    //
+    // Other expressions can convert an empty tensor into a non-empty tensor;
+    // particularly pad, cat, and reduction ops. These ops might have
+    // non-empty outputs so in order to guarantee that all (non- input or
+    // output) tensors are removed, we need to replace those ops with an
+    // equivalent that does not have any empty inputs.
+    for (auto use : tv->uses()) {
+      // If use is a ReductionOp or WelfordOp over some empty axes, replace it
+      // with a call to full().
+      if (auto rop = dynamic_cast<ReductionOp*>(use)) {
+        auto out = maybeReplaced(rop->out()->as<TensorView>());
+        if (hasEmptyRootReductionAxis(out)) {
```

Comment:
What happens if only some of zero domains are reduced? Should it be converted to a full op?

Reply:
In that case, the output will have some zero dims, so this reduction will be dead code.

Reply:
But the reduction itself is still replaced, and it's just it's going to be dead code eventually, right?

Reply:
Yes. I think I see what you're saying. It's unnecessary to replace the reduction if we know it's dead code. I can add a check for empty unreduced dims.